lianad/src/database/sqlite/utils.rs

```diff
@@ -307,6 +322,167 @@ fn migrate_v5_to_v6(conn: &mut rusqlite::Connection) -> Result<(), SqliteDbError
+                    Ok(tx)
+                },
+            )?;
+            db_exec(conn, |db_tx| {
+                let updated = db_tx.execute(
+                    "UPDATE transactions SET num_inputs = ?1, num_outputs = ?2, is_coinbase = ?3 WHERE txid = ?4",
+                    rusqlite::params![tx.input.len(), tx.output.len(), tx.is_coinbase(), txid[..].to_vec()],
+                )?;
+                assert_eq!(updated, 1);
+                Ok(())
+            })?;
+        }
+    }
+
+    // Update the `is_from_self` column for all unconfirmed coins and those
+    // confirmed after height 0, i.e. this will act on all coins.
+    let prev_tip_height = 0;
+    // As part of the same db_tx, first make sure that all rows of the
+    // transactions table have been updated.
+    db_exec(conn, |db_tx| {
+        let num_txs_to_update: u32 = db_tx.query_row(
+            "SELECT count(txid) FROM transactions WHERE num_inputs IS NULL OR num_outputs IS NULL",
+            [],
+            |row| row.get(0),
+        )?;
+        assert_eq!(num_txs_to_update, 0);
+
+        // This is a copy of `SqliteConn::update_coins_from_self` as of
+        // the time of writing this migration. We don't use that method
+        // directly in case the schema changes in future and it no longer
+        // works on the V7 schema.
+
+        // Given the requirement for unconfirmed coins that all ancestors
+        // be from self, we perform the update in a loop until no further
+        // rows are updated in order to iterate over the unconfirmed coins.
+        // Although we don't expect any unconfirmed transaction to have
+        // more than 25 in-mempool descendants including itself, there
+        // could be more descendants in the DB following a reorg and a
+        // rollback of the tip. The max number of iterations would be
+        // one per unconfirmed coin not from self plus one for all
+        // confirmed coins.
+        // In any case, the query only sets `is_from_self` to 1 for
+        // those coins with value 0 and so the number of rows affected
+        // by each iteration must become 0.
+        let max_iterations = {
+            let num_unconfirmed: u64 = db_tx.query_row(
+                "SELECT COUNT(*) FROM coins
+                WHERE blockheight IS NULL AND is_from_self = 0",
+                [],
+                |row| row.get(0),
+            )?;
+            // Add 1 for the confirmed coins, which will all
+            // be updated in the first iteration, and another 1
+            // as a final check there's nothing left to update.
+            num_unconfirmed.checked_add(2).expect("must fit")
+        };
+        let mut i = 0;
+        loop {
+            i += 1;
+            if i > max_iterations {
+                panic!(
+                    "More than {} iterations while updating is_from_self.",
+                    max_iterations
+                );
+            }
```

Comment:
I think you can directly do `for i in 0..max_iterations` with the updated == 0 as a break if it is finished early.

Reply:
I've changed it to a `for` loop and modified the logging slightly.